### ISSUE TITLE
Merged branches stay around until they're manually deleted

### ISSUE THREAD

**Issue by joshaber** [desktop/desktop#750]
@shiftkey writes

> In previous versions of Desktop, we had this behaviour for pruning:
> 
> - after a sync was completed, this is a chance to find branches that have been merged and removed locally
> - if the current branch is in a detached HEAD or unknown state, skip pruning
> - we look for branches which have been merged into the default branch (not the current branch)
> - on Windows we also had some additional guards to skip reserved branch names, like `gh-pages`
>
> I've created a branch [`prune-merged-braches-demo`](https://github.com/desktop/desktop/compare/prune-merged-braches-demo) that shows how you might add this into our process, but is deliberately incomplete - the `TODO` statements need to be addressed as part of implementing this.
>
> There might be some changes with how we do this, but this is a great chance for someone who cares about this feature to have a go at implementing it!

**Problem statement:**

Branches stick around in Desktop indefinitely until manually deleted. I want the branches that show up in Desktop to be relevant to me and not muddied by stale branches.

**Proposed solution:**

If 
- a branch is deleted on GitHub.com **and** 
- merged locally into the default branch in Desktop **and** 
- has no commits ahead of the default branch

then 
- they should be deleted in Desktop after a certain amount of time (to allow for people who frequently reuse the same branch for their work if that's a common workflow).

**Considerations:**
1. If there's a difference between the commits of the remote on GitHub.com and your local branch, we should not delete the local branch.
2. If GitHub has a concept of reserved branch names as @shiftkey references above (like `gh-pages`), we should skip pruning those branches.
3. If there's work that's **not** merged into default on GitHub.com (for example, a PR that's closed and deleted), we should not delete the local branch because it has work that's ahead of the default branch.

**Questions:**
1. How long before we automatically prune branches? (14 days)
2. How frequently do we check branches for ability to be pruned? (once per day)
3. Do we show anything in the UI to indicate that a branch will be pruned? (We don't think so - we're being intentionally conservative here and we don't think we'll put you in a position to lose any work here)
4. Are there reserved branches here that we should account for? (unknown)
5. Does it make sense to consider squashed and merged branches here (#1771)? (unknown)

**Comment by alrz** [comment]
Copying my comment from https://github.com/desktop/desktop/issues/1128#issuecomment-302165993 at @shiftkey request.

>when I choose to update all remotes it'd be nice if it run in parallel -- `git remote update --prune` does it sequentially.

**Comment by caiofbpa** [comment]
After I merge PRs on GitHub, they end up stale in my local repo. So I often run locally the following command:

```shell
git branch --merged | grep -v "master" | grep -v "*"| xargs git branch -d
```

I'm thinking about opening a PR that adds a "Prune merged branches" (needs better wording) to the Repository or Branch menus.

Is this what this issue is about? Or is it about pruning remote branches that were fetched locally?

**Comment by iAmWillShepherd** [comment]
@caiofbpa this is about pruning local branches. Your original thinking is correct.

**Comment by jpike88** [comment]
Ugh this issue is over a year old, too bad. I thought more people would be in need of something like this.

**Comment by shiftkey** [comment]
@jpike88 I've updated the issue with extra details about  the feature and what's required, to make clear that this is something an external contributor can pick up and have a go at!

**Comment by iAmWillShepherd** [comment]
@billygriffin could you clarify _which_ branches we should be pruning. Specifically, should I 
1. Prune all branches in all repos that Desktop tracks or 
1. Only prune the branches of the currently opened repo?

If we go with **option 1**, there's a chance the app could be closed mid-way through the cleanup, meaning we need to persist this state to some longer term place than memory. Do you have a preferred way of handling this? Do we resume when app opens or when the scheduled time arrives in situations like this?

If we go with *option 2**, that may lead to a scenario where the branches are never marked for cleanup because the user briefly opened the repo before moving on, not giving us enough time to kick off our cleanup process.

**Comment by billygriffin** [comment]
Good question @iAmWillShepherd.

> If we go with option 1, there's a chance the app could be closed mid-way through the cleanup, meaning we need to persist this state to some longer term place than memory.

If the app is closed mid-way through the cleanup, then the next time we check, wouldn't it just do the check again and anything that didn't get caught the first time would now get pruned appropriately? I'm not certain why we'd need to persist state. It doesn't really matter if it doesn't get all the way cleaned up if we'll be checking again at some interval, right?

**Comment by j-f1** [comment]
This should take into account the last time the branch was checked out/created. I can imagine some use cases where it might be useful to locally create a branch with an old revision where it would be annoying if the branch was deleted.

**Comment by iAmWillShepherd** [comment]
@j-f1, how much time do you think should pass before we step in and auto-clean it for you?

**Comment by j-f1** [comment]
I’m not really sure. Perhaps a couple of weeks?

**Comment by IdanAdar** [comment]
The changelog mentions that a merged or deleted branch will be pruned after 2 weeks. Why after 2 weeks? Why do I need to see the deleted branch in the dropdown for 2 extra weeks after it was deleted?

**Comment by billygriffin** [comment]
@IdanAdar Thanks for the feedback. We wanted to start conservatively on this since it’s a potentially destructive action and because it’s being newly introduced in the app - previously branches deleted on GitHub would remain forever until manually deleted in Desktop. The goal here is to make sure that the list doesn’t get so long with irrelevant branches that it becomes basically useless, and we think this is a good step toward that. We’re definitely open to making the logic more aggressive as we get feedback from folks, but we’re going to start here and see how things go and whether it’s alleviating the problem.

**Comment by tooomm** [comment]
I'm wondering if there a popup that warns you before deletion and offers you to keep certain branches?

**Comment by billygriffin** [comment]
@tooomm There isn't. We decided in taking a very conservative posture that we would only delete branches that aren't actively in use (hence the 14 days requirement) and don't have anything on them that's ahead of the default branch on GitHub. This combination, in addition to the branch being deleted on GitHub.com, seems relatively uncontroversial, but we're definitely going to listen to feedback we receive in the weeks and months after this ships to make sure it aligns with people's expectations and isn't overly aggressive.

**Comment by tooomm** [comment]
>we would only delete branches that aren't actively in use (hence the 14 days requirement) **and don't have anything on them that's ahead of the default branch on GitHub.**

I wasn't aware of the second part. 👍

**Comment by Kradirhamik** [comment]
Hi,

Sorry about my inexperience, but just to let you know that this issue is apparently still happening to me: when I delete a branch on the web, GitHub Desktop doesn't delete it - I have to do it manually every time.

Thanks!

**Comment by tierninho** [comment]
@Kradirhamik Stay tuned for branch pruning as it will be released in the coming weeks. Thank you for the patience.

**Comment by Croydon** [comment]
Is there a setting to disable this behaviour? 

While I see the potential advantages I still want to be in control when stuff gets deleted.

**Comment by billygriffin** [comment]
@Croydon Thanks for sharing your thoughts. Because we've taken an extremely conservative approach here, we don't intend to provide the ability to toggle it as a setting. If you use a particularly branch frequently, this will never prune that branch, and it's restricted only to branches where there is no additional work that hasn't been merged into the default branch on GitHub. In other words, we set this up intentionally such that users should never lose any work, which is our primary concern when deleting things. Hope that makes sense, and appreciate the feedback.

**Comment by GhostLyrics** [comment]
Is this restricted only to Github.com hosted repositories or is it actually:

- a branch is deleted on **the default remote** and
- merged locally into the default branch in Desktop and
- has no commits ahead of the default branch

I'm asking because I use the Desktop client as a general purpose git tool, not just solely for github.com projects.

**Comment by billygriffin** [comment]
@GhostLyrics Thanks for the question! For now it is restricted to GitHub.com. We chose to make this very conservative since it is destructive and we have a much better understanding of the information we receive from the GitHub API relative to other hosting providers. We likely won’t optimize for other services in the near term but we aren’t opposed to revisiting down the line.

**Comment by VelocityIsNotSpeed** [comment]
Do you check if there are stashed changes (or just changes) for the branch?

**Comment by outofambit** [comment]
@VelocityIsNotSpeed we don’t currently but we’ve talked about adding that logic. would you be willing to open a new issue for that?

**Comment by yay741** [comment]
[

](url)

**Comment by yay741** [comment]
Fix

**Comment by yay741** [comment]
[null](url)

**Comment by yay741** [comment]
We

**Comment by portal-thtcc1891993** [comment]
Law owner huy tuan tran tuan id , my

**Comment by zzj0402** [comment]
```bash
git remote prune origin
```

**Comment by yekota-copart** [comment]
still the issue is not resolved

**Comment by hrvojebusic** [comment]
This should be at least an option that user can opt-into for non-Github.com hosted repositories.

When working with non-GitHub hosted repositories with a lot of active contributors, pull becomes useless (always fails with an error) as many merged branches exist and pruning is required.

**Comment by JosephTLyons** [comment]
Any idea on what the progress of this is?

**Comment by billygriffin** [comment]
@JosephTLyons This is shipped, with the caveat that it's only for repos on GitHub.com. Due to feedback while we were building this, we decided to use 14 days as the constraint prior to pruning local branches so we're not pruning branches people were planning to reuse.

**Comment by JosephTLyons** [comment]
Thanks for the information, I didn't realize it had shipped.. most likely because I ended up manually deleting branches before 14 days.  Thanks for letting me know.

**Comment by TaxiTweedy** [comment]
> @JosephTLyons This is shipped, with the caveat that it's only for repos on GitHub.com. Due to feedback while we were building this, we decided to use 14 days as the constraint prior to pruning local branches so we're not pruning branches people were planning to reuse.

I understand why you would wait 14 days to prune, and i like the option to delete manually in the app once i've merged and deleted the branch on github.com. But....when manually deleting a branch in. the app, the pop up warning states that the branch still exists on remote. Is this a bug?

**Comment by billygriffin** [comment]
@TaxiTweedy Are you saying you're getting the popup even when the branch no longer exists on the remote? If so, I'm guessing this might be a particularly tricky condition where the app hasn't yet fetched and therefore isn't aware yet that the branch has been deleted.

**Comment by TaxiTweedy** [comment]
> @TaxiTweedy Are you saying you're getting the popup even when the branch no longer exists on the remote? If so, I'm guessing this might be a particularly tricky condition where the app hasn't yet fetched and therefore isn't aware yet that the branch has been deleted.

Yes, exactly. A branch was merged and deleted on remote but appears in the desktop app. This in itself i understand (pruning after 14 days) however the manual option to delete warns that it still exists. I waited several hours after deleting on remote before trying to delete (on app) to see if it was a fetch issue. It feels like the language on the pop up (saying branch exists on remote) was written as static and perhaps doesn't change even if it has been deleted from remote?

**Comment by tidy-dev** [comment]
@TaxiTweedy I was able to reproduce this behavior.  There are two ways to delete a branch: 1) via branch dropdown context menu and 2) via app branch menu. The app branch menu was displaying the correct dialog message, but the branch dropdown context menu was not. I have opened a PR to address this bug. Thank you for the report!

**Comment by TaxiTweedy** [comment]
> There are two ways to delete a branch: 1) via branch dropdown context menu and 2) via app branch menu.

I did not know this (and also just discovered the shortcut)! Thank you for the heads up!

**Comment by Gakk** [comment]
@billygriffin I am still having the same problem as @TaxiTweedy are reporting. After merging a pull request om GitHub.com i choose to delete the branch from the pull request interface. When returning to GitHub Desktop it shows me an action tip to publish the branch:

![image](https://user-images.githubusercontent.com/5013909/128697753-809d982a-a9d0-4232-a424-95bd7d8f0161.png)

I thought that GitHub Desktop should prune branches deleted from GitHub.com after pull request merge?

**Comment by shvetalg** [comment]
True @Garkk I am having the same issue...

**Comment by Slluxx** [comment]
Why is this after 6 years still not a thing? People that are just starting out feel like they will break something here. Our GitHub repositorys are set up to delete branches of merged PRs. That leaves us with a bunch of non existing branches that are still visible in Github Desktop. A small button "Sync branches with remote" would help tremendously.

**Comment by ekwoka** [comment]
Why is this issue closed when it was never addressed and other issues are pointed here as the place to follow it?

**Comment by Gakk** [comment]
I believe, @Slluxx and @ekwoka, that this feature actually is implemented. Scrolling through earlier comments in this issue I find:

> - We've taken an extremely conservative approach [...].
> - We chose to make this very conservative since it is destructive.

And from the changelog I find that the feature was implemented in [version 2.1.0](https://github.com/desktop/desktop/blob/release-2.1.0/changelog.json#L4) released in July 2019:

> [New] Branches that have been merged and deleted on GitHub.com will now be pruned after two weeks - #750"

Unfortunately there seems to be even another big limitation for my part, and that is that only branches and pull requests that use _normal merges_ will be pruned. If your workflow uses _squash and merge_ you are out of luck, as this is not detected.

Any improvements to this already implemented issue should probably be registered as a new issue.

**Comment by ekwoka** [comment]
Thanks. I tried to look through it but I couldn't get much that clarified if and how it worked (it was just clear it wasn't working).

2 weeks can be quite long but manageable, but yes, we do squash and merge everywhere, so it won't work anyway, sadly.

**Comment by jackbentley** [comment]
> we do squash and merge everywhere, so it won't work anyway, sadly.

These tools may be of use to you:
 - https://github.com/not-an-aardvark/git-delete-squashed
 - https://github.com/jackbentley/git-delete-gh-merged

**Comment by ekwoka** [comment]
Yes, I've been using `gh-poi` which does this.
